src/ILCompiler.Compiler/src/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -89,9 +89,20 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(startup.GetKnownMethod("GetMainMethodArguments", null)));
             }
             codeStream.Emit(ILOpcode.call, emitter.NewToken(_mainMethod));
+
+            // If the main method returns void, return the latched exit code set by Environment.ExitCode.
             if (_mainMethod.Signature.ReturnType.IsVoid)
             {
-                codeStream.EmitLdc(0);
+                MetadataType environmentAugments = Context.SystemModule.GetType("Internal.Runtime.Augments", "EnvironmentAugments");
```

Comment:
It may be better to hook this up via StartupCodeHelpers.

For ns20, we will also want to issue shutdown events here; or wait for background threads to finish. It would be nice to have one type where all the startup/shutdown helpers for compiler auto generated main live.